onnxruntime/core/optimizer/gemm_transpose_fusion.cc

```diff
@@ -26,29 +26,37 @@ Status GemmTransposeFusion::Apply(Graph& graph, Node& node, RewriteRuleEffect& m
 
   // check if input A is a Transpose
   if (A_node_ptr != nullptr && A_node_ptr->OpType() == "Transpose") {
-    Node& A_node = *graph.GetNode(A_node_ptr->Index());
-    transA = !transA;
-    if (A_node.GetOutputEdgesCount() > 1) {
-      // remove the edge between the Transpose and Gemm nodes
-      int output_idx = graph_utils::GetNodeOutputIndexFromOutputName(A_node, gemm_node.MutableInputDefs()[0]->Name());
-      graph.RemoveEdge(A_node.Index(), gemm_node.Index(), output_idx, 0);
-    } else {
-      nodes_to_remove.push_back(A_node);
+    // make sure all consumers are gemm nodes to avoid possible double transpose 
+    std::vector<const Node*> gemm_nodes = graph_utils::FindChildrenByType(*A_node_ptr, "Gemm");
+    if (gemm_nodes.size() == A_node_ptr->GetOutputEdgesCount()) {
+      Node& A_node = *graph.GetNode(A_node_ptr->Index());
+      transA = !transA;
+      if (A_node.GetOutputEdgesCount() > 1) {
+        // remove the edge between the Transpose and Gemm nodes
+        int output_idx = graph_utils::GetNodeOutputIndexFromOutputName(A_node, gemm_node.MutableInputDefs()[0]->Name());
+        graph.RemoveEdge(A_node.Index(), gemm_node.Index(), output_idx, 0);
+      } else {
+        nodes_to_remove.push_back(A_node);
+      }
+      new_gemm_input_defs[0] = A_node.MutableInputDefs()[0];
     }
-    new_gemm_input_defs[0] = A_node.MutableInputDefs()[0];
   }
   // check if input B is a Transpose
   if (B_node_ptr != nullptr && B_node_ptr->OpType() == "Transpose") {
-    Node& B_node = *graph.GetNode(B_node_ptr->Index());
-    transB = !transB;
-    if (B_node.GetOutputEdgesCount() > 1) {
-      // remove the edge between the Transpose and Gemm nodes
-      int output_idx = graph_utils::GetNodeOutputIndexFromOutputName(B_node, gemm_node.MutableInputDefs()[1]->Name());
-      graph.RemoveEdge(B_node.Index(), gemm_node.Index(), output_idx, 1);
-    } else {
-      nodes_to_remove.push_back(B_node);
+    // make sure all consumers are gemm nodes to avoid possible double transpose 
+    std::vector<const Node*> gemm_nodes = graph_utils::FindChildrenByType(*B_node_ptr, "Gemm");
+    if (gemm_nodes.size() == B_node_ptr->GetOutputEdgesCount()) {
+      Node& B_node = *graph.GetNode(B_node_ptr->Index());
+      transB = !transB;
+      if (B_node.GetOutputEdgesCount() > 1) {
+        // remove the edge between the Transpose and Gemm nodes
```

Comment:
can you add a further comment here explaining why we just remove the edge and how the transpose node removal will happen for the last gemm this is triggered on?